highway/crud_object.go

```diff
@@ -2,22 +2,158 @@ package highway
+			ObjectDoc: &ot.ObjectDoc{
+				Label:       res.WhatIs.ObjectDoc.GetLabel(),
+				Description: res.WhatIs.ObjectDoc.GetDescription(),
+				Did:         res.WhatIs.ObjectDoc.GetDid(),
+				BucketDid:   res.WhatIs.ObjectDoc.GetBucketDid(),
+				Fields:      outputFields,
+			},
+			Creator:   res.WhatIs.GetCreator(),
+			Timestamp: res.WhatIs.GetTimestamp(),
+			IsActive:  res.WhatIs.GetIsActive(),
+		},
+	}, nil
 }
 
 // UpdateObject updates an object.
 func (s *HighwayServer) UpdateObject(ctx context.Context, req *ot.MsgUpdateObject) (*ot.MsgUpdateObjectResponse, error) {
-	return nil, ErrMethodUnimplemented
+	// Verify that there are requested changes
+	if req.GetAddedFields() == nil && req.GetRemovedFields() == nil {
+		return nil, errors.New("object to register must have fields")
+	}
+
+	// Translate ot fields to otv1
+	addedFields := s.bufToBlockchain(req.GetAddedFields())
+	removedFields := s.bufToBlockchain(req.GetRemovedFields())
+
+	// Build Transaction
+	tx := &otv1.MsgUpdateObject{
+		Creator:       req.GetCreator(),
+		Label:         req.GetLabel(),
+		AddedFields:   addedFields,
+		RemovedFields: removedFields,
+	}
+
+	// Broadcast the message
+	res, err := s.cosmos.BroadcastUpdateObject(tx)
+	if err != nil {
+		return nil, err
+	}
+
+	// decode otv1 fields to ot
+	outputFields := s.decodeObjectDocFields(res.WhatIs.ObjectDoc.Fields)
+
+	return &ot.MsgUpdateObjectResponse{
+		Code:    res.GetCode(),
+		Message: res.GetMessage(),
+		WhatIs: &ot.WhatIs{
+			Did: res.WhatIs.GetDid(),
+			ObjectDoc: &ot.ObjectDoc{
+				Label:       res.WhatIs.ObjectDoc.GetLabel(),
+				Description: res.WhatIs.ObjectDoc.GetDescription(),
+				Did:         res.WhatIs.ObjectDoc.GetDid(),
+				BucketDid:   res.WhatIs.ObjectDoc.GetBucketDid(),
+				Fields:      outputFields,
+			},
+			Creator:   res.WhatIs.GetCreator(),
+			Timestamp: res.WhatIs.GetTimestamp(),
+			IsActive:  res.WhatIs.GetIsActive(),
+		},
+	}, nil
 }
 
-// // DeleteObject deletes an object.
-// func (s *HighwayServer) DeleteObject(ctx context.Context, req *ot.MsgDeleteObject) (*ot.MsgDeleteObjectResponse, error) {
-// 	return nil, ErrMethodUnimplemented
-// }
+// DeleteObject deletes an object.
+func (s *HighwayServer) DeactivateObject(ctx context.Context, req *ot.MsgDeactivateObject) (*ot.MsgDeactivateObjectResponse, error) {
+	// Build Transaction
+	tx := &otv1.MsgDeactivateObject{
+		Creator: req.GetCreator(),
+		Did:     req.GetDid(),
+	}
+
+	// Broadcast the message
+	res, err := s.cosmos.BroadcastDeactivateObject(tx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ot.MsgDeactivateObjectResponse{
+		Code:    res.GetCode(),
+		Message: res.GetMessage(),
+	}, nil
+}
+
+// -----------------
+// Helper Functions
+// -----------------
+
+// Translate ot (buf) fields to otv1 (blockchain) //TODO rename this
+func (s *HighwayServer) bufToBlockchain(bufFields []*ot.ObjectField) []*otv1.ObjectField {
+	var blockchainFields []*otv1.ObjectField
+	for _, v := range bufFields {
```

Comment:
we should add some checks on the collection for before iterating over it.

Reply:
nill pointer check and length check

Reply:
I believe these would both be redundant with the range function, it will simply not iterate on nil or 0 len.

Reply:
we should guard so we can provide contextual information, perhaps it should return ([]*ot_v1.ObjectField, error)